docs/reference/c++_support.rst

```diff
@@ -0,0 +1,165 @@
+
+All C++20 language features are supported, but extensions and restrictions apply. C++20
+introduced features that change programs foundationally with coroutines and modules, but
+unfortunately these aren't supported by HIP. On the other hand, ``consteval`` functions
+can be called from host and device, even if it is specified for host use only. 
+
+The three-way comparison operator (also known as spaceship operator ``<=>``) works in
+both host and device code.
+
+.. _language_restrictions:
+Extensions and Restrictions
+===============================================================================
+
+Besides the above mentioned deviations from the standard, there are more general
+extensions and restrictions to consider. 
+
+``__global__`` functions
+-------------------------------------------------------------------------------
+
+Functions that work as an entry point for device execution are called kernels and are
+specified with the ``__global__`` qualifier. Calling these functions happen with the
+triple chevron operator: ``<<< >>>``. Kernel functions return type must be ``void``, they
+can't be ``constexpr``, can't have a parameter of type ``std::initializer_list`` or
+``va_list``, can't have a parameter of rvalue reference type. Kernels can have variadic
+template parameters, but only one pack, and it must be the last in the template parameter
+list. 
+
+Device Space Memory Specifiers
+-------------------------------------------------------------------------------
+
+To specify whether a variable is allocated in host or device memory, HIP has qualifiers
+called device space memory specifiers. These are ``__device__``, ``__shared__``,
+``__managed__`` and ``__constant__``. These are meant to specify what memory is allocated
+for a variable. ``__device__`` and ``__constant__`` specifier are used for static
+variables and they will be allocated on the HIP devices global memory, with the only
+difference that the ``__constant__`` variables can't be changed from the device code.
+``__shared__`` is used in device code, to allocate the variable on the shared memory,
+which is available for all threads in a block. ``__managed__`` is also used on global
+variables. A managed variable is emitted as an undefined global symbol in the device
+binary and is registered by ``__hipRegisterManagedVariable`` in init functions. The HIP
+runtime allocates managed memory and uses it to define the symbol when loading the device
+binary. A managed variable can be accessed in both device and host code. It is important
+to know where each variable is because they will be only available in certain places.
+Generally, variables allocated in the host memory will not be available from device code
+and variables allocated in the device memory will not be available from host code. This
+can be an issue mostly with pointers in host memory, which are pointing to device memory.
+Dereferencing these will cause a segmentation fault.
```

Comment:
HIP includes device space memory specifiers to indicate whether a variable is allocated in host or device memory and how its memory should be allocated. HIP supports the `` __device__``, ``__shared__``, ``__managed__`` and ``__constant__`` specifiers. 

The ``__device__`` and ``__constant__`` specifiers define static variables, which are allocated within global memory on the HIP devices. The only difference is that ``__constant__`` variables can’t be changed after allocation. The ``__shared__`` specifier allocates the variable within shared memory, which is available for all threads in a block.

The ``__managed__`` variable specifier is used to create global variables that are initially undefined and unaddressed within the global symbol table. The HIP runtime allocates managed memory and defines the symbol when it loads the device binary. A managed variable can be accessed in both device and host code. To register managed variables, use ``__hipRegisterManagedVariable`` in an init function.

It’s important to know where a variable is stored because it is only available from certain locations. Generally, variables allocated in the host memory are not accessible from the device code, while variables allocated in the device memory are not accessible from the host code. Dereferencing a pointer to device memory on the host results in a segmentation fault.

Reply:
If init is the name of a function, then use inline code formatting. Otherwise, spell it out as initialization. If you mean a constructor, change this to object constructor.